File: src/main/java/com/akagiyui/drive/model/LoginUserDetails.java
package com.akagiyui.drive.model;

import com.akagiyui.drive.entity.Role;
import com.akagiyui.drive.entity.User;
import com.fasterxml.jackson.annotation.JsonIgnore;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;
import org.springframework.security.core.GrantedAuthority;
import org.springframework.security.core.authority.SimpleGrantedAuthority;
import org.springframework.security.core.userdetails.UserDetails;

import java.io.Serializable;
import java.util.Collection;
import java.util.List;
import java.util.Objects;
import java.util.Set;
import java.util.stream.Collectors;

/**
 * 登录用户详情
 * @author AkagiYui
 */
@Data
@AllArgsConstructor
@NoArgsConstructor
public class LoginUserDetails implements UserDetails, Serializable {

    /**
     * 用户
     */
    User user;

    /**
     * 权限字符串
     */
    private Set<String> permissions;

    /**
     * 权限列表
     */
    @JsonIgnore
    private Set<GrantedAuthority> authorities;

    /**
     * 登录用户详情
     * @param user 用户
     */
    public LoginUserDetails(User user) {
        this.user = user;

        // 获取权限，注意如果使用懒加载，需要在事务内获取
        List<Role> roles = user.getRoles();
        this.permissions = roles.stream()
                .map(Role::getPermissions)
                .flatMap(Collection::stream)
                .map(Permission::name)
                .collect(Collectors.toSet());
    }

    /**
     * 获取用户权限
     * @return 用户权限集合
     */
    @Override
    public Collection<? extends GrantedAuthority> getAuthorities() {
        // 如果 authorities 为空，则初始化，缓存加速
        return Objects.requireNonNullElseGet(authorities, () -> {
            // 转换为 GrantedAuthority 集合
            authorities = permissions.stream()
                    .map(SimpleGrantedAuthority::new)
                    .collect(Collectors.toSet());
            return authorities;
        });
    }

    /**
     * 获取用户密码
     * @return 用户密码
     */
    @Override
    @JsonIgnore
    public String getPassword() {
        return user.getPassword();
    }

    /**
     * 获取用户名
     * @return 用户名
     */
    @Override
    @JsonIgnore
    public String getUsername() {
        return user.getUsername();
    }

    /**
     * 用户账号是否未过期
     * @return 用户账号是否未过期
     */
    @Override
    @JsonIgnore
    public boolean isAccountNonExpired() {
        return true;
    }

    /**
     * 用户账号是否未被锁定
     * @return 用户账号是否未被锁定
     */
    @Override
    @JsonIgnore
    public boolean isAccountNonLocked() {
        return true;
    }

    /**
     * 用户密码是否未过期
     * @return 用户密码是否未过期
     */
    @Override
    @JsonIgnore
    public boolean isCredentialsNonExpired() {
        return true;
    }

    /**
     * 用户是否可用
     * @return 用户是否可用
     */
    @Override
    @JsonIgnore
    public boolean isEnabled() {
        return !user.getDisabled();
    }
}

File: src/main/kotlin/com/akagiyui/drive/model/LoginUserDetails.kt
package com.akagiyui.drive.model

import com.akagiyui.drive.entity.User
import com.fasterxml.jackson.annotation.JsonIgnore
import org.springframework.security.core.GrantedAuthority
import org.springframework.security.core.authority.SimpleGrantedAuthority
import org.springframework.security.core.userdetails.UserDetails

/**
 * 登录用户详情
 * @author AkagiYui
 */
class LoginUserDetails(
    /**
     * 用户
     */
    var user: User
) : UserDetails {

    /**
     * 权限字符串
     */
    lateinit var permissions: Set<String>

    /**
     * 权限列表
     */
    private val innerAuthorities: MutableSet<out GrantedAuthority> by lazy {
        permissions
            .map { SimpleGrantedAuthority(it) }
            .toMutableSet()
    }

    /**
     * 登录用户详情
     */
    init {
        // 获取权限，注意如果使用懒加载，需要在事务内获取
        val roles = user.roles
        this.permissions = roles
            .map { it.permissions }
            .flatten()
            .map { it.name }
            .toSet()
    }

    /**
     * 获取用户权限
     * @return 用户权限集合
     */
    @JsonIgnore
    override fun getAuthorities() = innerAuthorities

    /**
     * 获取用户密码
     * @return 用户密码
     */
    @JsonIgnore
    override fun getPassword(): String = user.password

    /**
     * 获取用户名
     * @return 用户名
     */
    @JsonIgnore
    override fun getUsername(): String = user.username

    /**
     * 用户账号是否未过期
     * @return 用户账号是否未过期
     */
    @JsonIgnore
    override fun isAccountNonExpired() = true

    /**
     * 用户账号是否未被锁定
     * @return 用户账号是否未被锁定
     */
    @JsonIgnore
    override fun isAccountNonLocked() = true

    /**
     * 用户密码是否未过期
     * @return 用户密码是否未过期
     */
    @JsonIgnore
    override fun isCredentialsNonExpired() = true

    /**
     * 用户是否可用
     * @return 用户是否可用
     */
    @JsonIgnore
    override fun isEnabled(): Boolean = !user.disabled

}
